### DISCUSSION TITLE
Is it possible to update firmware without the app

### DISCUSSION THREAD

**Discussion by DutchessNicole** [Anders-Holst/xled_plus#22] [Q&A] [answered]
Hey.

I just bought a set of Twinkly LED strips and would like to add them to HA, however that requires a firmware version I don't have yet. The current firmware version on the device is 2.6.6.

The app uses a discovery protocol of some kind that does not work, because the Twinkly controller is on a separate subnet with client isolation on. It's meant for IoT stuff so it's pretty much locked down completely.

Since I do know the IP address of the device, can I get the firmware from somewhere and manually upgrade it on the device using the xled library?

**Comment by Anders-Holst** [comment, marked as answer]
I do not know actually.

The xled package has a function for updating the firmware, and I essentially copied it over to xled_plus, but I have never tried it myself, since I have no access to any firmware files.

I can see several reasons why you want to do it the way you propose, for example if you are uncertain about the latest version and want to go back to a previous version if something has stopped working. However, the simplest solution in your case is probably to try to download it through the app anyway: Change the device network mode to ap-mode, connect to its wavelan on your phone while making sure you have data traffic on, and then start the app and let it install the latest firmware for you. Then you can set it back to station-mode connected to your secure network. (Provided the LED strip has the same protocol and capabilities as the LED strings.)

(By the way, if it is on a secure subnet, how would it be able to communicate with HA? Isn't the point to protect it from outside communication? But network stuff is not my area, so maybe I am mistaken...)

**Reply by DutchessNicole** [reply]
A secure subnet in this case means that all intra-device communication is disabled (as far as client isolation can do that, there's ways around that) and no devices can connect to the internet at large. I make exceptions for devices that absolutely MUST connect to something important outside, like NTP for devices that refuse my internal NTP advertisement.

All devices can, however, connect to their router on their network segment. And HA lives on a separate network segment that has unrestricted access to the IoT network.

So in essence, HA can access anything in the IoT network through simple routing, and replies can go back to HA through the TCP established session filter, but no IoT device can connect to anywhere but its router unless I allow it.

**Reply by DutchessNicole** [reply]
This is also how I would like the app to work: it should run on my general 'trusted devices' network and connect by IP directly to the Twinkly controller in the IoT network. That connection is allowed. However the app does not allow me to select an IP, it must broadcast apparently and broadcasts are not routed, so the controller never gets the signal and therefore doesn't answer.

**Reply by Anders-Holst** [reply]
I see. Then both xled and xled_plus may come in handy, since you can do almost everything with them that you can do with the app. (again, provided that it is the same protocol as the strings.) Once you have downloaded the latest firmware, that is. (Is there such a thing as "layout mapping" for strips? Then you may have to do that while on the less restricted network too, because it needs the app.)

Did you try AP mode and connect directly to it with the phone? An alternative is to temporarily connect it to a less restricted network while downloading the firmware, and then move it back. (Note that if you use xled/xled_plus to switch between the networks you may have to use `set_network_mode_station_v2` (where "v2" is the significant part) since the old function is for older firmwares than yours.)
